File: backend/src/main/scala/domains/submission/objects/request/CreateSubmissionRequest.scala
package domains.submission.objects.request

import domains.submission.objects.*

import domains.problem.objects.ProblemSlug
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

/** 创建提交的请求体；programs 按角色名保存语言和源码。 */
final case class CreateSubmissionRequest(
  problemSlug: ProblemSlug,
  programs: Map[String, CreateSubmissionRequest.Program]
)

/** CreateSubmissionRequest 的 JSON 编解码器和嵌套程序结构。 */
object CreateSubmissionRequest:
  /** 单个提交角色的语言和源码。 */
  final case class Program(
    language: SubmissionLanguage,
    sourceCode: SubmissionSourceCode
  )

  /** Program 的 JSON 编解码器。 */
  object Program:
    given Encoder[Program] = deriveEncoder[Program]
    given Decoder[Program] = deriveDecoder[Program]

  given Encoder[CreateSubmissionRequest] = deriveEncoder[CreateSubmissionRequest]
  given Decoder[CreateSubmissionRequest] = deriveDecoder[CreateSubmissionRequest]

File: frontend/src/objects/submission/request/CreateSubmissionRequest.ts
import type { ProblemSlug } from '@/objects/problem/ProblemSlug'
import type { SubmissionLanguage } from '@/objects/submission/SubmissionLanguage'
import type { SubmissionSourceCode } from '@/objects/submission/SubmissionSourceCode'

/** 创建提交请求中的程序定义；key 由 Record 外层提供。 */
type CreateSubmissionProgram = {
  language: SubmissionLanguage
  sourceCode: SubmissionSourceCode
}

/** 创建提交请求体；支持多程序提交，programs 的 key 由前端表单定义。 */
export type CreateSubmissionRequest = {
  problemSlug: ProblemSlug
  programs: Record<string, CreateSubmissionProgram>
}
